ray-operator/controllers/ray/rayservice_controller.go

```diff
@@ -1113,47 +1108,34 @@ func (r *RayServiceReconciler) reconcileServe(ctx context.Context, rayServiceIns
 			} else {
 				logger.Info("Skipping the update of Serve applications because the Ray head Pod is not ready.")
 			}
-			return false, err
+			return false, serveApplications, nil
 		}
 	}
 
+	var clientURL string
 	if clientURL, err = utils.FetchHeadServiceURL(ctx, r.Client, rayClusterInstance, utils.DashboardPortName); err != nil || clientURL == "" {
-		return false, err
+		return false, serveApplications, err
 	}
 
 	rayDashboardClient := r.dashboardClientFunc()
 	if err := rayDashboardClient.InitClient(ctx, clientURL, rayClusterInstance); err != nil {
-		return false, err
+		return false, serveApplications, err
 	}
 
 	cachedServeConfigV2 := r.getServeConfigFromCache(rayServiceInstance, rayClusterInstance.Name)
-	isReady, newApplications, err := getAndCheckServeStatus(ctx, rayDashboardClient)
-	// TODO (kevin85421): We should unify the update of the CR status in calculateStatus.
-	rayServiceStatus.Applications = newApplications
+	isReady, serveApplications, err := getAndCheckServeStatus(ctx, rayDashboardClient)
 	if err != nil {
-		return false, err
+		return false, serveApplications, err
 	}
-	shouldUpdate, reason := checkIfNeedSubmitServeApplications(cachedServeConfigV2, rayServiceInstance.Spec.ServeConfigV2, newApplications)
+	shouldUpdate, reason := checkIfNeedSubmitServeApplications(cachedServeConfigV2, rayServiceInstance.Spec.ServeConfigV2, serveApplications)
 	logger.Info("checkIfNeedSubmitServeApplications", "shouldUpdate", shouldUpdate, "reason", reason)
 
 	if shouldUpdate {
 		if err = r.updateServeDeployment(ctx, rayServiceInstance, rayDashboardClient, rayClusterInstance.Name); err != nil {
-			return false, err
+			return false, serveApplications, err
 		}
 	}
-
-	logger.Info("Check serve health", "isReady", isReady, "isActive", isActive)
-
-	if !isReady {
-		// TODO (kevin85421): avoid always updating status if the serve applications are not ready.
-		rayServiceInstance.Status.ServiceStatus = rayv1.WaitForServeDeploymentReady
-		if err := r.Status().Update(ctx, rayServiceInstance); err != nil {
```

Comment:
Return the serve applications and pass them to `calculateStatus` to avoid calling `Status().Update(...)` here.